### PR TITLE
6188 subscription not marked as a cross partition subscription matches operation on resources in other partitions

### PR DESCRIPTION
**Pull request by epeartree**
What was done:
- provided a failing test;
- modified the criterias to consider a subscription to be a cross-partition subscription.  A Subscription will considered a cross-partition subscription if:

  * global flag subscription.cross_partition.enable = TRUE;
       AND
  * the Subscription is created on the DEFAULT partition;
       AND
  * the Subscription has extension `https://smilecdr.com/fhir/ns/StructureDefinition/subscription-cross-partition ` = TRUE;
- renamed/removed tests;
- added changelog

### CODE REVIEW

File: hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/CanonicalSubscriptionTest.java [hunk 1]
```diff
@@ -86,40 +89,48 @@ public void emailDetailsEquals() {
...
-		subscription.addExtension(HapiExtensions.EXTENSION_SUBSCRIPTION_CROSS_PARTITION, new BooleanType().setValue(false));
+		subscription.setUserData(RESOURCE_PARTITION_ID,theRequestPartitionId);
+		subscription.addExtension(HapiExtensions.EXTENSION_SUBSCRIPTION_CROSS_PARTITION, new BooleanType().setValue(true));
+
 		CanonicalSubscription canonicalSubscription = canonicalizer.canonicalize(subscription);
 
 		System.out.print(canonicalSubscription);
 
-		assertEquals(canonicalSubscription.getCrossPartitionEnabled(), theIsCrossPartitionEnabled);
+		// for s Subscription to be a cross-partition subscription, u need 3 things:
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/CanonicalSubscriptionTest.java:129]:
```suggestion
		// for a Subscription to be a cross-partition subscription,  3 things are required:
```

File: hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/CanonicalSubscriptionTest.java [hunk 2]
```diff
@@ -86,40 +89,48 @@ public void emailDetailsEquals() {
...
+		// - The Subs need to be created on the default partition
+		// - The Subs need to have extension EXTENSION_SUBSCRIPTION_CROSS_PARTITION set to true
+		// - Global flag CrossPartitionSubscriptionEnabled needs to be true
+		assertThat(canonicalSubscription.isCrossPartitionEnabled()).isEqualTo(theExpectedIsCrossPartitionEnabled);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/test/java/ca/uhn/fhir/jpa/subscription/module/CanonicalSubscriptionTest.java:133]: nitpick: you have listed 3 things, but have only made 1 assert. Is that sufficient?

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java [hunk 3]
```diff
@@ -781,11 +780,23 @@ public String getSubscriptionStatus(IBaseResource theSubscription) {
...
 	}
 
-	private void handleCrossPartition(IBaseResource theSubscription, CanonicalSubscription retVal) {
-		if (mySubscriptionSettings.isCrossPartitionSubscriptionEnabled()) {
-			retVal.setCrossPartitionEnabled(true);
-		} else {
-			retVal.setCrossPartitionEnabled(SubscriptionUtil.isCrossPartition(theSubscription));
+	private boolean handleCrossPartition(IBaseResource theSubscription) {
+		RequestPartitionId requestPartitionId =
+				(RequestPartitionId) theSubscription.getUserData(Constants.RESOURCE_PARTITION_ID);
```
Comment (tadgh) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:785]: question: Are you sure this userdata key is set in all cases where partitioning is involved? e.g. tenant,patient-id, and manual mode?

**Review by tadgh — APPROVED** [review]
Approved pending various questions.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java (hunk 3 above)
Comment (epeartree) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionCanonicalizer.java:785]: Key RESOURCE_PARTITION_ID is not always present in the userData map.  in such case, the invocation returns null which is checked at line 789 before calling method isDefaultPartition() on the object.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_4_0/6188-non-cross-partition-subs-match-all-resources.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 6188
+jira: SMILE-8759
+title: "Previously, a Subscription not marked as a cross-partition subscription matched operation on resources in other
```
[the remark this answers is not in the document]
Comment (epeartree) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_4_0/6188-non-cross-partition-subs-match-all-resources.yaml:5]: modified

**Pull request merged**